### PR TITLE
[RFR] Defer loading

### PR DESCRIPTION
**Pull request by psav**
* Many cases where we were comparing to a mgmt_system object, instead of the provider object itself
* Deferred the loading of mgmtsystem classes until needed
* Moved some deployment args into a deployment_helper function, making the code in utils generic

### CODE REVIEW

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

File: cfme/infrastructure/provider/virtualcenter.py [hunk 1]
```diff
@@ -1,10 +1,11 @@
-from mgmtsystem.virtualcenter import VMWareSystem
 from . import InfraProvider
+from mgmtsystem.virtualcenter import VMWareSystem
 
 
 @InfraProvider.add_provider_type
 class VMwareProvider(InfraProvider):
     type_name = "virtualcenter"
+
```
Comment (mshriver) [inline cfme/infrastructure/provider/virtualcenter.py]: no mgmt_class property change here?

File: utils/virtual_machines.py [hunk 2]
```diff
@@ -60,52 +52,30 @@ def deploy_template(provider_key, vm_name, template_name=None, timeout=900,
...
         try:
-            deploy_args.update(template=data['small_template'])
+            deploy_args.update(template=provider_crud.data['small_template'])
         except KeyError:
             raise ValueError('small_template not defined for Provider {} in cfme_data.yaml'.format(
                 provider_key))
     else:
         deploy_args.update(template=template_name)
 
-    if isinstance(mgmt, RHEVMSystem):
```
Comment (mshriver) [inline utils/virtual_machines.py]: ah - good to get rid of this in favor of the deployment_helper

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

File: cfme/infrastructure/provider/virtualcenter.py (hunk 1 above)
Comment (psav) [inline cfme/infrastructure/provider/virtualcenter.py]: yeh I musta missed it

File: cfme/cloud/provider/azure.py [hunk 3]
```diff
@@ -1,12 +1,15 @@
...
-from mgmtsystem.azure import AzureSystem
 from . import CloudProvider
 
 
 @CloudProvider.add_provider_type
 class AzureProvider(CloudProvider):
     type_name = "azure"
-    mgmt_class = AzureSystem
+
+    @property
```
Comment (jkrocil) [inline cfme/cloud/provider/azure.py]: All these can be `@cached_property`.

File: cfme/tests/control/test_actions.py [hunk 4]
```diff
@@ -92,7 +92,7 @@ def pytest_generate_tests(metafunc):
     pytest.mark.meta(blockers=[
         BZ(
             1149128,
-            unblock=lambda provider: not isinstance(provider.mgmt, mgmtsystem.scvmm.SCVMMSystem))
+            unblock=lambda provider: not isinstance(provider, SCVMMProvider))
```
Comment (jkrocil) [inline cfme/tests/control/test_actions.py]: `not provider.one_of(SCVMMProvider)`

File: cfme/tests/control/test_actions.py [hunk 5]
```diff
@@ -631,7 +631,7 @@ def test_action_initiate_smartstate_analysis(
         test_flag: actions, provision
     """
     # Set host credentials for VMWare
-    if isinstance(vm.provider, mgmtsystem.virtualcenter.VMWareSystem):
+    if isinstance(vm.provider, VMwareProvider):
```
Comment (jkrocil) [inline cfme/tests/control/test_actions.py]: `if vm.provider.one_of(VMwareProvider):`

File: cfme/tests/infrastructure/test_provisioning.py [hunk 6]
```diff
@@ -19,7 +19,7 @@
     pytest.mark.meta(blockers=[
         BZ(
             1265466,
-            unblock=lambda provider: not isinstance(provider.mgmt, RHEVMSystem))
+            unblock=lambda provider: not isinstance(provider, RHEVMProvider))
```
Comment (jkrocil) [inline cfme/tests/infrastructure/test_provisioning.py]: Same as above; I'm not going to comment on all the instances.

File: utils/virtual_machines.py [hunk 7]
```diff
@@ -60,52 +52,30 @@ def deploy_template(provider_key, vm_name, template_name=None, timeout=900,
...
-    elif isinstance(mgmt, GoogleCloudSystem):
-        pass
-    elif isinstance(mgmt, AzureSystem):
-        deploy_args.update(data['provisioning'])
-    elif isinstance(mgmt, OpenstackSystem):
-        if ('network_name' not in deploy_args) and data.get('network'):
-            deploy_args.update(network_name=data['network'])
-    else:
-        raise Exception("Unsupported provider type: {}".format(type(mgmt).__name__))
+    deploy_args.update(provider_crud.deployment_helper(deploy_args))
```
Comment (jkrocil) [inline utils/virtual_machines.py]: :+1:

**Pull request merged**